### DISCUSSION TITLE
MiniJV880: Rompler synth emulation

### DISCUSSION THREAD

**Discussion by probonopd** [probonopd/MiniDexed#726] [Show and tell]
https://github.com/giulioz/mini-jv880 - @giulioz has performed a heart surgery and transplanted another synth engine!

> MiniJV880 is a rompler-style synthesizer closely modeled on the famous JV-880 by a well-known Japanese manufacturer running on a bare metal Raspberry Pi (without a Linux kernel or operating system).

https://www.youtube.com/watch?v=7LWaYHTNIDI

**Comment by probonopd** [comment]
Now, wouldn't it be cool if we could multiboot this from the same SD card? Might be possible by using [`os_prefix`](https://www.raspberrypi.com/documentation/computers/config_txt.html#os_prefix) in `config.txt`.

**Reply by plamikcho** [reply]
Very cool project! Is it compatible with MiniDexed hardware? What about the configs? There is no information in this repo.

**Reply by giulioz** [reply]
Hey! Author here. So far I developed the thing just to work on my hardware setup, but it should work on other setups with some very minimal changes. I'm trying to find some free time to work on it and I will write some docs soon!

**Reply by blixabloxa** [reply]
Been playing with this using the nuked sc55 emulator. Would be great to have it co-installed on the minidexed SD card!

**Reply by penfold42** [reply]
> Now, wouldn't it be cool if we could multiboot this from the same SD card? Might be possible by using [`os_prefix`](https://www.raspberrypi.com/documentation/computers/config_txt.html#os_prefix) in `config.txt`.

combined with a  filter depending on the state of a GPIO. For example:
```
[gpio4=1]
# Settings here apply if GPIO 4 is high
```
Could let you have a front panel switch to select minijv or minidexed at boot/powerup.

The application files from  https://github.com/Sterr1/Mini-JV880pi and minidexed don't seem to clash so should co-exist.

I've just done a quick test with a combined SD card that only has the application files and bootcode.bin on it. 
The "kernel" and related files are all pxe/tftp booted from the network and I can flip a symbolic link to switch between ubuntu, minijv and minidexed.

**Comment by miotislucifugis** [comment]
This looks super!   I second the request for making hardware configs compatible with minidexed (serial and i2c 1602 displays and assignable GPIOs for the buttons and encoder.)   i wouldnt even mind if I had to swap SD cards, but multiboot, whoa!

**Comment by plamikcho** [comment]
Hi, I made some attempts to get the MiniJV880 working on the same hardware as MiniDexed with the help of the author @giulioz (huge thanks). The code is here: https://github.com/plamikcho/mini-jv880/tree/codespace-humble-goldfish-r79xv4x557hp4jx The good news are that it is working and it is possible to get it running. It requires more buttons (13 more + the rotary encoder button) but the Pi has enough free GPIOs. My changes in @giulioz 's code are quite naive, mostly adapted code from MiniDexed because I do not have experience in C++. The problems I've hit (and cannot solve due to lack of knowledge):
- LCD display is truncated, SSD1306 supports 20 columns but JV880 has 24. My solution is just to cut some of the columns for each row, I've skipped the places with spaces or brackets. This is not optimal and the other menus are broken. To get all of the 24 columns it needs to be rendered by pixels and not chars which requires direct operation with SSD1306 + scaling the JV880 display to 128x32 (this is already done by the author)
- LCD display lacks cursor (SSD1306 issue, in MiniDexed the cursor is not working too). Unfortunately JV880 uses the cursor to cycle through items operated with the rotary encoder. (update, found a way to fix it, now it has a blinking cursor)
- Buttons are not responsive, maybe the debounce time needs some adjustment (or my code has some mistake which I cannot figure out) - this was probably a consequence from the missing cursor and the bad display, now buttons work fine.
- The sound is distorted and has glitches - this probably comes from the multicore adaptation of the emulator for RPi, again no idea how to fix. Recordings are made on RPi3 which performed better (???), RPi4 sound was terrible with distortion and glitches (quite strange for me, I expected the better hardware to work smooth).

JV880 emulator is really cool and sounds really nice, it would be great to run together with MiniDexed (changing the cards or dual boot). I hope that the experienced developers here can help to finish it and to enjoy the sounds on the affordable MiniDexed hardware.
https://audio.com/plamen-dzorev/audio/jv880-test-dowah
https://audio.com/plamen-dzorev/audio/jv880-test-trumpet
https://audio.com/plamen-dzorev/audio/jv880-test-jazzorgan
![IMG_20241020_150445](https://github.com/user-attachments/assets/77a4d9a1-1515-4898-82ea-1e1ece50da42)
UPDATE: Again with the help of the author I have some playable patches on RPi4 processing the emulator on a single core. Finally enjoying a nice sounding sax on the MiniDexed :)
https://audio.com/plamen-dzorev/audio/jv880test-turboa
Fixed the buttons and the display cursor. Removed the spinlocks from pcm. First 5-10 mins before the Pi gets hot the patches sound good. Need to add cooling fan.
![IMG_20241025_143624](https://github.com/user-attachments/assets/2179707e-c62e-4110-ab30-5dc123fd26e3)

**Reply by temporal-distortion** [reply]
Very great work. Silly question - will it allow for using a wave expansion (sr-jv)?

**Reply by franbriz** [reply]
Hi @plamikcho !! You are a f*ck|ng genius... I don't know how you can compile the project. And make it work in RPI4. I'm trying to compile for a RPI Zero 2. But it is imposible for me. Can you help me? I'm beginner using github. I was clone the repository with dependences and I tried for 2 days compiling the original project for RPI 1 and 2. But always take an error. If you can tell me a little "instruction" or anything, I would be very grateful. Thanks!! (sorry for my english).

**Reply by plamikcho** [reply]
I just used the instructions for Minidexed here: https://github.com/probonopd/MiniDexed/wiki/Development#developing-in-the-cloud. I couldn't setup local environment, it always ended with an error on my mac. For RPi Zero 2 you should change the code as it was previously to process on 2 cores (it is commented in my branch as I remember). This worked with minimum crackling and glitches (you can see @giulioz video on youtube)

**Comment by temporal-distortion** [comment]
I have quite some issue with MiniJV880 and an audio-hat from Raspiaudio, but for some odd reason not with the official Iqaudio+/raspberry dac+ audio hat:

![Issue_with_audioDAC](https://github.com/user-attachments/assets/50242266-af9a-41dc-a7c1-301ab0944eb4)

Tried them both on rPI4 as well as rPI5. Minidexed works always, but miniJV880 shows garbled characters when using the Raspiaudio hat (but always works with the iqAudio dac+).

**Comment by Sterr1** [comment]
Working prototype [here](https://github.com/Sterr1/Mini-JV880pi)

**Reply by temporal-distortion** [reply]
Hi @Sterr1 (and of course @giulioz giulioz

Idk the "why" this is, but while i can successfully run and play both miniDexed (and dreamdexed fork), the corresponding ini-file setting(with regards to midi, audio and LCD) does not work with the miniJV880. Sometimes when i boot the 880, the whole thing freezes. Rebooting it most often mean i can at least scroll through the patches, but i can't hear anything.

The above on a rPI5 with a domoshop midi hat (https://domoshop.eu/products/slim-midi-shield-for-raspberry-pi-3-4) and an IQAudio DAC+ hat where mini and dreamdexed boots and works 100% fine every time i boot them.

I did setup a minimalistic setup using a rPI3 with an LCD and rotary encoder (using USB midi) and i did get sound, it was only that the sound was extremely fragmented (i heard a few stutters of the patch for a second before it sounded as it should).

On my rPI setup, i did boot it with Raspberry OS again just to verify I2S addresses (in case they would have drifted apart somehow) but they kept being constant:

DAC+ on 0x4c
LCD on 0x27
Midi on 0x21


![jv880](https://github.com/user-attachments/assets/bdaf6a65-5a38-4c3b-8766-117c0c65a43c)


Any suggestions what to try?


My minijv880.ini file for reference


#
# minijv880.ini
#

# Sound device
SoundDevice=i2s
#SoundDevice=pwm
#SoundDevice=hdmi
SampleRate=32000
#ChunkSize=256
DACI2CAddress=0x4c
ChannelsSwapped=0
# Engine Type ( 1=Modern ; 2=Mark I ; 3=OPL )
EngineType=1

# Master Volume (0-127)
MasterVolume=100

# MIDI
MIDIBaudRate=31250
#MIDIThru=umidi1,ttyS1
IgnoreAllNotesOff=0
MIDIAutoVoiceDumpOnPC=0
HeaderlessSysExVoices=0
# Program Change enable
#   0 = Ignore all Program Change messages.
#   1 = Respond to Program Change messages.
MIDIRXProgramChange=1
# Program Change mode
#   0 = Only recognise Program Change 0-31.
#   1 = Support 0-127 across four consecutive banks.
# NB: Only relevant if PerformanceSelectChannel=0
ExpandPCAcrossBanks=1
# Program Change action:
#   0 = Program Change messages select voices on the channel associated with each TG.
#   1-16 = Program Change messages on this channel select performances.
#   >16 = Program Change messages on ANY channel select performances.
# NB: In performance mode, all Program Change messages on other channels are ignored.
PerformanceSelectChannel=0


# Default is 16x2 display (e.g. HD44780)
# Any LCD present ?
LCDEnabled=1
LCDColumns=24
LCDRows=2
LCDI2CAddress=0x27

# GPIO Button Navigation
#  Any buttons set to 0 will be ignored
ButtonPinPreview=
ButtonActionPreview=click
ButtonPinLeft=
ButtonActionLeft=click
ButtonPinRight=
ButtonActionRight=click
ButtonPinData=
ButtonActionData=click
ButtonPinToneSelect=
ButtonActionToneSelect=click
ButtonPinPatchPerform=
ButtonActionPatchPerform=click
ButtonPinEdit=
ButtonActionEdit=click
ButtonPinSystem=
ButtonActionSystem=click
ButtonPinRhythm=
ButtonActionRhythm=
ButtonPinUtility=
ButtonActionUtility=click
ButtonPinMute=
ButtonActionMute=
ButtonPinMonitor=
ButtonActionMonitor=
ButtonPinCompare=
ButtonActionCompare=
ButtonPinEnter=11
ButtonActionEnter=click

# Timeouts in milliseconds for double click and long press
DoubleClickTimeout=400
LongPressTimeout=400

# KY-040 Rotary Encoder
EncoderEnabled=1
EncoderPinClock=10
EncoderPinData=9

# Debug
MIDIDumpEnabled=0
ProfileEnabled=0

**Comment by Sterr1** [comment]
You probably haven't tried my release. I recompiled the non-working source code from giulios and plamikcho, which were terribly buggy, and created a working release that doesn't click or produce harmonics because I've slightly modified the pcm code. Of course, I compiled it for my rpi zero 2w, but it should be compatible with rpi 3. Everything sounds fine to me. In the latest release, I've added serial midi. And one person has already tried it and reported that it works. If you have some bugs, write issue on my project page. Rpi 5 still not supported.

**Reply by temporal-distortion** [reply]
Kudos Sterr1, I'll try it out on a rPI3 (with said IQAdio DAC++) and the move on to the rPI5 from there (same setup). Much appreciated.

**Reply by Sterr1** [reply]
Rpi 5 in my release not supported.

**Reply by plamikcho** [reply]
Tested on RPi4 - Sterr1's release is much better than the original code, thanks! Some patches still has clicks and glitches but now there are only few of them.

**Reply by Sterr1** [reply]
there is one realtime parameter that is desirable to adjust, depends on the processor speed. It should be moved to the config, but I didn't have time yet. I have configured this parameter on rpi2w, on other processors with a different frequency, it needs to be changed.

**Reply by temporal-distortion** [reply]
I tried your latest build (1.0.3) on my rPI4 (coupled to my HD44780 20x2 display and the IQAudio+ DAC) using a good 16GB card. Looking at the HDMI output, i see this and it looks ok (i guess)

![IMG_20250828_114436](https://github.com/user-attachments/assets/52d53f07-77d9-47c0-b494-92bb71cd7c23)

But while the LCD lights up, nothing shows up on it.

if i take the same card and pop either minidexed or dreamdexed onto it and boot the rPI4 up, i get corresponding HDMI output but also the nice menu on the LCD.

**Reply by Sterr1** [reply]
Build only for 2w/3. Other not guaranteed.

**Reply by temporal-distortion** [reply]
Got you :)

It works perfect on the rPI3 (just checked)

**Comment by blixabloxa** [comment]
Love the possibilities here. I am wondering if a Cheap Yellow Device (CYD) screen could be incorporated into this or even MiniDexed to provide patch info and touchscreen buttons and controls for mini-jv880, instead of a physical knob and buttons. Will require some serious coding if possible.

**Comment by Maximo68** [comment]
Ciao, io ho un raspberry Pi4b, esiste una release da provare sul proprio hardware?

**Reply by Sterr1** [reply]
No, check releases in project page. May be in future. Or you can compile kernel for rpi4 from source.

**Reply by Maximo68** [reply]
Non ho idea di come si compila... E' possibile avere una foto di come è composta la Directory di File e Cartelle nella scheda SD. 
Ho scaricato questa: https://github.com/Sterr1/Mini-JV880pi/releases/download/1.0.3/Mini-JV880pi.serial.zip ...     Ma il display rimane acceso senza alcuna informazione. Manca qualche file? Grazie x il vostro lavoro.

**Reply by Sterr1** [reply]
Rpi4 still not supported. Another driver sd card used. May be next release.

**Comment by miotislucifugis** [comment]
be sure enable the proper screen (and disable* the other ones) in the 880.ini file.  otherwise, the display will not work.  (Had me stumped for a moment- the top row of my 1602 serial display was all solid blocks...)
 *I think I had to change a i2c address to 0x00.

**Reply by Sterr1** [reply]
No. I think Rpi4 can't read sd card after boot.

**Reply by temporal-distortion** [reply]
This is my impression as well. On both rPI5 and rPI4 (both non-supported) i also got this MBR issue preventing it from booting. On the 3, it worked flawlessly :)

(here is me hoping/fingers crossed for a 4/5 release in the future as the 5 (double fingers crossed) supports external power on/off, etc which is important for me :)

**Reply by Maximo68** [reply]
Confermo che ora funziona! Tutto Ok su rPI4b. Ho eliminato gli schermi non necessari e mantenuto solo quello in uso un HD44780 LCD con indirizzo LCDI2CAddress=0x27

**Reply by Maximo68** [reply]
![1756493555595](https://github.com/user-attachments/assets/5328afd5-85b2-4437-b35a-304233af7ebe)

**Reply by temporal-distortion** [reply]
Maximo68, this is super cool. Do you mean to say that by disabling all the other LCD definitions, except the one that you use, you actually got it to work for you?

Which fork did you use in that case?

**Reply by Maximo68** [reply]
Scusa puoi riformulare la domanda? Devo tradurre dall'inglese all'italiano e non capisco il senso della tua richiesta...

**Reply by temporal-distortion** [reply]
Certo, nessun problema. Che build hai usato per farlo funzionare su un Raspberry PI4? E l'audio funziona?

Lo chiedo perché non sono mai riuscito a farlo funzionare su un Raspberry 5 (non si avvia, niente audio) e con Raspberry 4 a volte sono riuscito ad avviarlo, ma non c'era alcun audio.

Su Raspberry 3, funziona tutto.

Purtroppo, nel mio caso, mi affido a un pulsante di accensione esterno, USB-C, ecc. (che solo rPI5 può fornire).

(sorry for the bad translation Maximo :))

**Reply by Sterr1** [reply]
Not understand. English please

**Reply by temporal-distortion** [reply]
:) My comment to Maximo:

Sure, no problem. What build did you use to get it to run on a Raspberry PI4, and does the sound work as well?

I'm asking because i have never been able to get it to work on a Raspberry 5 (Its not booting, no sound) and with Raspberry 4 i did sometimes get it to boot, but there was no sound at all. 

On Raspberry 3, everything works.

Unfortunately, in my case, i rely on an external power-button, USB-C (compared to rPI3 micro-usb) etc (that only rPI5 would provide)

In my case, i'm eager to learn as i built this 19u rack around an rPI5 with an original intention of using miniDexed (later on dreamDexed) but when i found out about the virtualJV, i was completely hooked on the concept. I choose rPI5 in this case as it allows for a discrete power on/off button (NO) and in the miniDexed case, some upgraded polyphony (32 voices instead of 24 on rpi4 or 16 on rpi3).

![IMG-20250712-214421](https://github.com/user-attachments/assets/40d01795-0ce0-4d7a-a4c4-018258e556f3)

**Reply by Maximo68** [reply]
Ho usato questa build. https://github.com/Sterr1/Mini-JV880pi/releases/download/1.0.3/Mini-JV880pi.serial.zip
L'audio funziona perfettamente. Ora devo capire come importare nuovi preset e se è necessario una modifica hardware con pulsanti aggiuntivi per far funzionare il tutto.

**Reply by temporal-distortion** [reply]
Bene :) Quindi sei riuscito a farlo funzionare sull'rPI 4?

L'ho provato su un rPI 4 (IQAudio + display I2C 20x2) e il suono era estremamente discontinuo (o ritardato. Dopo aver premuto i tasti, c'era questo ritardo di 1 secondo prima che il suono iniziasse a essere riprodotto).

Ho provato ora a evidenziare tutto ciò che non era correlato all'I2C e riesco a far funzionare lo schermo, ma ancora niente suono :(

-----------------------

Nice :) So you got it to work on the rPI 4?

I tried it on a rPI 4 (IQAudio + 20x2 I2C display) and the sound was extremely choppy (or delayed. After hitting the keys, there was this 1 second delay before the sound started to play).

I tried now to remark everything not related to I2C, and i get the screen working but still no sound :(

**Comment by Maximo68** [comment]
Potrebbe essere un collegamento al Dac interrotto o un' impostazione errata delle specifiche del file minijv880.ini ?
Io sto usando l'Interfaccia I2S PCM5102A e non ho mai avuto problemi. Anche il Minidexed funziona alla grande.

**Reply by temporal-distortion** [reply]
Immagino che tutto sia possibile, ma posso cambiare la scheda SD e caricare Minidexed/Dreamdexed in qualsiasi momento e funziona (usando l'IQAudio DAC+, che è anche la scheda DAC ufficiale del Raspberry, se non sbaglio, su rPI3 ma non su rPI4 o rPI5).

Ho anche la scheda PCM5102A (a dire il vero non l'ho ancora provata), ma non capisco perché quella dovrebbe funzionare mentre quella "ufficiale" no.

Continuerò a seguire gli sviluppi. Con un po' di fortuna, anche rPI5 sarà supportato (se Minidexed/Circle lo supporta ora, spero che anche l'emulazione 880 faccia lo stesso in futuro, dato che si basa sulla stessa base di codice :))


------------------------

Anything is possible i guess, but i can constantly swap the SD-card and load Minidexed/dreamdexed and it just works for me (using the IQAudio DAC+ which is also the Raspberry DAC official board if I'm not mistaken, on, on the rPI3 but not on the rPI4 or rPI5).

I also have the PCM5102A pcb (haven't tried it out yet to be honest) but  i cant understand why that one should work where the "official" one would not?

I will keep following any development here. With some luck, even the rPI5 will be supported eventually (if minidexed/Circle supports it now, i hope that also the 880 emulation will do the same in the future as it builds on the same code-base :))

**Comment by Sterr1** [comment]
[NEW](https://github.com/Sterr1/Mini-JV880pi/discussions) Now full display of JV880 shown.

**Comment by oldmaga** [comment]
Hi everyone,
after some hard work... this is my version of the miniJV-880:
Rpi4
2402 I2C display
PCM5102 DAC
KY-040 Rotary Encoder
GPIO keyboard (thanks to plamickho for the idea...)
MIDI Shield

![a40be6ac-4125-4bb5-b5a4-8cff3118848b](https://github.com/user-attachments/assets/61837fe9-5856-4077-acbe-1cd796da1641)
![fc567b92-dbe0-4f06-bdb4-dd0cfa977b77](https://github.com/user-attachments/assets/69a23ca1-85ae-45cc-a208-fa588a04feac)
![52894df4-e08f-433e-92fb-c8c0ed0beea7 (1)](https://github.com/user-attachments/assets/a1bde319-86e8-4fca-984e-d1f3128e2e77)
Everything works great, only the GPIO button panel has some sporadic problems.
Of course it is perfectly compatible with the MiniDexed, just change the SD on the front.

Thanks to all of you for the great work done!

ps: just a clarification but I want to underline that my name has nothing to do with the idiotic MAGA movement!

**Reply by giulioz** [reply]
Very cool!
Would be very curious to see a video of it in action :)

**Reply by Sterr1** [reply]
What version of mini-jv880 did you use?

**Reply by oldmaga** [reply]
Your 1.0.6 version. 
I'm still having problems with the GPIO buttons (for the moment I can't use MIDI commands): when I press one button it works, but on the second press it doesn't, until I press another... I know it's a known issue and I hope it can be solved somehow

**Reply by Sterr1** [reply]
I'm aware of this issue. The problem is, I didn't change plamikcho code for handling the  buttons. MIDI buttons work by calling physical buttons events. I was informed about this bug after the release, but I was busy, so I didn't fix it. The next version will fix everything, and there will be many significant changes.  But don't you think it's better to discuss the Mini-JV880pi project not on the MiniDexed page?

**Comment by temporal-distortion** [comment]
oldmaga, you wouldn't happen to have the STL files you used here & the button matrix info to share by any chance?

It looks amazing and i would definately not mind trying to replicate this sweet device of yours :)

**Reply by oldmaga** [reply]
Sure, no problem.
I can provide you both STL files and information on the components I used, and perhaps we can continue the discussion on the MiniJV-880 site. I'll  soon give you the link ;)

**Reply by oldmaga** [reply]
Ok, files are now loaded on my account.
 Let me know if there are any problems

**Reply by temporal-distortion** [reply]
Nice. Thx oldmaga. Having a peek now.

**Reply by oldmaga** [reply]
Hi everyone,
with the precious help of some of you, after a few months of work I have finally finished the development and improvements to my MiniJV880. Looking at the photo you probably won't notice anything, because all the new features are under the hood.

First of all the (very recent) optional coexistence on the same SD of the mythy Minidexed (or DreamDexed...) and MiniJV880: dualboot is managed by holding down a button when turning on. I'm definitely not the first to do this but I'd say it works well. The button problems have been completely solved, and the encoder is now precise and does not miss beats. Furthermore, after a lot of sweat, access to the Ram Card, both for reading and writing, was also unlocked. 
<img width="3946" height="1671" alt="Image-01" src="https://github.com/user-attachments/assets/7489cf9c-a9d3-4dfa-a0f6-cc7f66ae4ff5" />
<img width="3978" height="2147" alt="Image-03" src="https://github.com/user-attachments/assets/ad480fab-ea7f-4284-841c-a8f293b39431" />


You can use the SR expansions and the PN-JV80 cards, via custom menus, and as a bonus also those of the RD-500 (as in VirtualJV). The .SYX files are organized in folders that can be completely managed directly from the panel or via Ethernet (Wi-Fi is not yet stable) using the embedded HTML (mini)server, from which it is possible to carry out many system management/maintenance operations, including kernel upload and reboot. I also added a serial/mini-usb interface for debugging/remote control.
<img width="1796" height="496" alt="Image-04-A" src="https://github.com/user-attachments/assets/54be5bcd-8bee-477a-af9a-3f0da64cad8a" />
<img width="1781" height="515" alt="Image-05-A" src="https://github.com/user-attachments/assets/5f51458b-71f2-4df7-8056-73980fb86b05" />
<img width="1875" height="560" alt="Image-08-A" src="https://github.com/user-attachments/assets/28e0fe42-6727-46ed-9dd2-6259a39b6c28" />
<img width="1781" height="521" alt="Image-10-A" src="https://github.com/user-attachments/assets/4cd2ae80-e40d-4b91-be58-f59825f95978" />


While I was at it, I created a series of PC-tools for managing the CardRam and for transferring files via TFTP, plus a "Remote GUI" for remote control, with two pushbutton/displays: that of the MiniJV880 and that of the real JV880, albeit with some arrangements.
<img width="1501" height="964" alt="ER_GUI" src="https://github.com/user-attachments/assets/c1898731-511d-4784-a037-9c55a2f7444c" />
<img width="1250" height="964" alt="CardRam Manager" src="https://github.com/user-attachments/assets/afefcc66-ae4f-476e-ba00-f7a37b87ecd6" />



I have to say, having both worlds on the same machine is a really satisfying thing.
If you're interested check it out, I'd really appreciate some of your feedback. Your criticisms, comments and suggestions will be welcome.
https://github.com/oldmaga/MiniJV880-CardRAM-public

Thanks again everyone for the help you gave me!

**Reply by giulioz** [reply]
@oldmaga incredibly cool!

**Reply by probonopd** [reply]
> dualboot is managed by holding down a button when turning on

Do you have a pointer for concise instructions on how to do this? I'd like to add it to our wiki. Thanks!

**Reply by oldmaga** [reply]
Hi, yes. I have a concise example here:

https://github.com/oldmaga/MiniJV880-CardRAM-public/releases/tag/v2.4.0-public-clean.10

The relevant file is:

src/config-dualboot-data-gpio12.txt

The idea is deliberately very simple: it is not a custom bootloader.
It uses the Raspberry Pi firmware `config.txt` GPIO conditional filters.

In my current setup, tested only on Raspberry Pi 4:

- both projects share the same SD-card boot partition;
- common Raspberry Pi boot files stay in the SD-card root;
- MiniJV880 kernel is placed in:

      /minijv880/kernel8-rpi4.img

- MiniDexed kernel is placed in:

      /minidexed/kernel8-rpi4.img

- GPIO12 is configured as input with pull-up;
- default/high state boots MiniJV880;
- holding the DATA button at power-on pulls GPIO12 low and boots MiniDexed.

Minimal relevant `config.txt` part:

    `[all]
    boot_delay=0
    disable_splash=1
    force_eeprom_read=0
    gpu_mem=16
    disable_overscan=0

    # GPIO12 high/default: MiniJV880
    # GPIO12 low/DATA held: MiniDexed
    gpio=12=ip,pu

    [pi4]
    arm_64bit=1
    armstub=armstub8-rpi4.bin

    # Default boot target
    kernel=/minijv880/kernel8-rpi4.img

    # DATA held at power-on: GPIO12 low -> MiniDexed
    [gpio12=0]
    kernel=/minidexed/kernel8-rpi4.img

    [all]`

Of course, the GPIO pin can be replaced with any other suitable free GPIO pin, and the button logic can be reversed if desired. I chose GPIO12 (DATA, according to my minijv880.ini), which in my case is a kind of "all-purpose" button.

In my MiniJV880 firmware I also made kernel maintenance dualboot-aware: MiniJV880 HTTP/TFTP update functions manage only the MiniJV880 kernel under `/minijv880/`. The MiniDexed kernel under `/minidexed/` is treated as read-only/detection-only, so it cannot be overwritten accidentally from the MiniJV880 maintenance workflow.

The wiki version could probably describe it as:

“Single FAT boot partition, two kernel subdirectories, Raspberry Pi `config.txt` GPIO filter selects which bare-metal kernel to start.”

Technical background: Raspberry Pi supports both the gpio=... directive in config.txt to set pull-up/input/output at boot, as well as conditional filters like [gpio4=1] and [gpio2=0]. In my public file, the actual example actually uses /minijv880/kernel8-rpi4.img, /minidexed/kernel8-rpi4.img, gpio=12=ip,pu, and override [gpio12=0].

**Reply by oldmaga** [reply]
I forgot: this obviously allows for more than two choices, just use additional GPIOs.

I'm thinking of trying to create a small bare-metal kernel selector, so as to have a real menu on the LCD/encoder: would that be interesting?

**Reply by Sterr1** [reply]
Yes. I was try, but not lucky. And by midi command please. :)

**Reply by penfold42** [reply]
Thanks for sharing - I'm literally doing this tomorrow with my new build.
Ideally we'd improve minidexed and minijv80 to look for their config/data files in separate top level directory. There no are no conflicts right now, but thats just luck!
Then we could use the gpio switching to set os_prefix and they could all live in self contained folders

**Reply by probonopd** [reply]
Thanks @oldmaga, I added it to https://github.com/probonopd/MiniDexed/wiki/Multiboot. Feel free to edit that wiki page if you like.

**Comment by my03max** [comment]
Wow!

I have most of the BOM in the checkout basket, but now is the time where i push the buy button :-D

**Comment by Maximo68** [comment]
Thanks everyone for making this update available, which seems to be working for now. I'd like to ask for clarification on the GPIO12 configuration as a pull-up input. Should the button be connected directly to ground with one terminal and the other to GPIO12? Or is a resistor of some value needed? Could you provide a small circuit diagram? Thanks again for your work.

**Reply by oldmaga** [reply]
All buttons must be connected the same way: one terminal connected to ground (use one of the 8 GND pins of the GPIO connector) and the other to the GPIO pin. So for the 12-button keyboard you only need 13 wires: 12 for the GPIO pins and one (connecting all the buttons) to a GND terminal. No resistance needed.

<img width="601" height="800" alt="Schema 12 pulsanti" src="https://github.com/user-attachments/assets/0e39ea94-f36f-46ab-862e-f07096c7dc34" />

If you need other information, just ask. If they also concern the MiniDexed we can stay here, but if they only concern the MiniJV880 I suggest you continue on my repo. Have fun!
